base/ltpara.dtx

```diff
@@ -0,0 +1,1147 @@
+%    conventions: \cs{@@@@par} and \cs{endgraf} both refer to the
+%    default meaning (in the past this was the initex primitive) while
+%    \cs{par} is the current meaning which may does something else.
+%
+%
+%    We are now going to change this default meaning to run
+%    \cs{para_end:} instead, which ultimately executes the initex
+%    primitive but additionally adds our hooks when appropriate.
+%    This way the change is again transparent to the legacy \LaTeXe{}
+%    code.
+%
+%    In most cases \cs{para_end:} should behave exactly like the
+%    primitive and we achieve this by simply expanding it to the
+%    primitive which is available to us as \cs{tex_par:D}. This way we
+%    don't have to care about whether \TeX{} just does nothing (e.g.,
+%    if in vertical mode already) or generate an error, etc.
+%    \begin{macrocode}
+\cs_new_protected:Npn \para_end: {
+%    \end{macrocode}
+%
+%    The only case we care about is when we are in horizontal mode
+%    (i.e., doing typesetting) and not also in inner mode (i.e.,
+%    making paragraphs and not building an \cs{hbox}.
+%    \begin{macrocode}
+%  \bool_lazy_and:nnT
+%       { \mode_if_horizontal_p: }
+%       { \bool_not_p:n { \mode_if_inner_p: } }
+%       { ...
+%    \end{macrocode}
+%    This is inner loop so we code the above test with low-level \cs{if_mode...}s.
+%    \begin{macrocode}
+  \if_mode_horizontal:
+    \if_mode_inner: \else:
+%    \end{macrocode}
+%    In that case the action of the primitive would be to remove the
+%    last glue (not kern) from the horizontal list constructed to form
+%    a paragraph then append the a penalty of 10000 and the
+%    \cs{parfillskip} at the end and pass the whole list to the
+%    paragraph builder which breaks it into lines and \TeX{} then
+%    returns to vertical mode.
+%
+%    What we want to do instead is to add our hook code at the end of
+%    the horizontal list before that happens and the code is passed to
+%    the paragraph builder. If there was a glue item at the end then
+%    it should get removed before the hook code gets added so we have
+%    to arrange for its removal ourselves.
+%
+%    There is not much point in checking if there was really a glue
+%    item at the end of the horizontal list, instead we simply try to
+%    remove one using \cs{tex_unskip:D}, if there wasn't one this will
+%    do nothing.
+%    \begin{macrocode}
+         \tex_unskip:D
+%    \end{macrocode}
+%    The we execute the public hook (which may add final typesetting
+%    material) followed by the kernel hook we need for adding tagging
+%    support. None of this is supposed to change the mode---at the
+%    moment we make only a very simple test for this, more devious
+%      changes go unnoticed, but too bad, that will then probably
+%      badly backfire.
+%    \begin{macrocode}
+         \hook_use:n{para/end}
+         \@kernel@after@para@end
+         \if_mode_horizontal:
```

Comment:
This `\if_mode_horizontal:` would need changing as well because of `para/after`

Reply:
hmm, that `para/hook` could be more of a problem, so I guess you are right and I should use the higher-level interfaces